src/snmalloc/ds_core/helpers.h

```diff
@@ -361,28 +339,46 @@ namespace snmalloc
       append(static_cast<unsigned long long>(x));
     }
 
-  public:
     /**
-     * Constructor.  Takes a format string and the arguments to output.
+     * Append with format string and arguments. Compiler
+     * is able to optimize the recursion into loops.
      */
-    template<typename... Args>
-    SNMALLOC_FAST_PATH MessageBuilder(const char* fmt, Args... args)
+    template<typename Head, typename... Tail>
+    void append_args(const char* fmt, Head&& head, Tail&&... tail)
```

Comment:
If you added 
```C++
    void append_args(const char* fmt)
    {
      append(fmt);
    }
```

Then you could use overloading to remove the
```C++
        if constexpr (sizeof...(Tail) == 0)
        {
          append(fmt + 2);
        }
        else
        {
          append_args(fmt + 2, std::forward<Tail>(tail)...);
        }
```
and just replace it with
```C++
        append_args(fmt + 2, std::forward<Tail>(tail)...);
```

Reply:
I also wonder if you could drop the special case:
```C++
    SNMALLOC_FAST_PATH MessageBuilder(const char* fmt)
    {
      buffer[SafeLength] = 0;
      for (const char* s = fmt; *s != 0; ++s)
      {
        append_char(*s);
      }
      append_char('\0');
    }
```
with this change?